api/src/services/user.service.ts

```diff
@@ -180,53 +182,85 @@ export class UserService {
-          throw new UnprocessableEntityException('No password for this user, was this account created via a provider?');
-        }
+    // Handle too many login attempts
+    if (loginAttempts >= 3) {
+      user.lastLogin = new Date();
+      await this.saveUser(user); // Save the updated lastLogin
+      throw new TooManyRequestsException('You have made too many requests. Please try again later.');
+    }
 
-        const valid = await new Promise((resolve, reject) => {
-          bcrypt.compare(password, user.encryptedPassword.toString('utf8'), (err, same) => {
-            if (err) {
-              reject(err);
-            } else {
-              resolve(same);
-            }
-          });
-        });
-        // When credentials are invalid, increment login attempts and respond with error
-        if (!valid) {
-          await this.userRepo.increment({ id: user.id }, 'loginAttempts', 1);
-          throw new UnauthorizedException('invalid credentials');
-        }
-        break;
+    // Handle password grant type authentication
+    if (grantType === GrantType.Password) {
+      await this.handlePasswordAuthentication(user, password, loginAttempts, userKey);
+    } else {
+      throw new BadRequestException('Tried to authenticate with unsupported grant type');
+    }
+
+    // All good, reset login attempts
+    await this.resetLoginAttempts(userKey, user);
+
+    return user;
+  }
+
+  private async getLoginAttempts(userKey: string, user: User): Promise<number> {
+    if (this.loginAttemptsStorage.getRedisClient()) {
+      // Fetch login attempts from Redis if available
+      return await this.loginAttemptsStorage.getUserAttempts(userKey);
+    }
+    // Fall back to using the database
+    return user.loginAttempts;
+  }
+
+  private async handlePasswordAuthentication(user: User, password: string, loginAttempts: number, userKey: string): Promise<void> {
+    if (!user.encryptedPassword) {
+      await this.incrementLoginAttempts(user, loginAttempts, userKey);
+      throw new UnprocessableEntityException('No password for this user. Was this account created via a provider?');
+    }
 
-      default:
-        throw new BadRequestException('Tried to authenticate with unsupported grant type');
+    const isValidPassword = await bcrypt.compare(password, user.encryptedPassword.toString('utf8'));
+
+    if (!isValidPassword) {
+      await this.incrementLoginAttempts(user, loginAttempts, userKey);
+      throw new UnauthorizedException('invalid credentials');
     }
+  }
 
+  private async incrementLoginAttempts(user: User, loginAttempts: number, userKey: string): Promise<void> {
     user.lastLogin = new Date();
+    await this.saveUser(user);
 
-    // All good, reset login attempts
-    user.loginAttempts = 0;
+    if (this.loginAttemptsStorage.getRedisClient()) {
+      await this.loginAttemptsStorage.setUserAttempts(userKey, loginAttempts + 1, 900); // 15 minutes TTL
+    } else {
```

Comment:
@samuelmbabhazi I don't like this "else" statement here honestly... Even if we are using Redis, I would still prefer to keep updating DB field of user (increment counter) so we can see how many times user tried to login and failed directly in main DB without need to go to redis to check that info!